kernel/src/table_changes/scan_file.rs

```diff
@@ -0,0 +1,447 @@
+    Ok(visitor.context)
+}
+
+// add some visitor magic for engines
+#[allow(unused)]
+struct CDFScanFileVisitor<'a, T> {
+    callback: CDFScanCallback<T>,
+    selection_vector: &'a [bool],
+    context: T,
+}
+
+impl<T> RowVisitor for CDFScanFileVisitor<'_, T> {
+    fn visit<'a>(&mut self, row_count: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<()> {
+        require!(
+            getters.len() == 18,
+            Error::InternalError(format!(
+                "Wrong number of CDFScanFileVisitor getters: {}",
+                getters.len()
+            ))
+        );
+        for row_index in 0..row_count {
+            if !self.selection_vector[row_index] {
+                continue;
+            }
+            let timestamp = getters[16].get(row_index, "scanFile.timestamp")?;
+            let commit_version: i64 = getters[17].get(row_index, "scanFile.commit_version")?;
+            let commit_version: u64 = commit_version.try_into().map_err(Error::generic)?;
+
+            if let Some(path) = getters[0].get_opt(row_index, "scanFile.add.path")? {
+                let deletion_vector = visit_deletion_vector_at(row_index, &getters[1..=5])?;
+                let dv_info = DvInfo { deletion_vector };
+                let partition_values =
+                    getters[6].get(row_index, "scanFile.add.fileConstantValues.partitionValues")?;
+                let scan_file = CDFScanFile {
+                    scan_type: CDFScanFileType::Add,
+                    path,
+                    dv_info,
+                    partition_values,
+                    commit_version,
+                    timestamp,
+                };
+                (self.callback)(&mut self.context, scan_file)
+            } else if let Some(path) = getters[7].get_opt(row_index, "scanFile.remove.path")? {
+                let deletion_vector = visit_deletion_vector_at(row_index, &getters[8..=12])?;
+                let dv_info = DvInfo { deletion_vector };
+                let partition_values = getters[13].get(
+                    row_index,
+                    "scanFile.remove.fileConstantValues.partitionValues",
+                )?;
+                let scan_file = CDFScanFile {
+                    scan_type: CDFScanFileType::Remove,
+                    path,
+                    dv_info,
+                    partition_values,
+                    commit_version,
+                    timestamp,
+                };
+                (self.callback)(&mut self.context, scan_file)
+            } else if let Some(path) = getters[14].get_opt(row_index, "scanFile.cdc.path")? {
+                let partition_values = getters[15]
+                    .get(row_index, "scanFile.cdc.fileConstantValues.partitionValues")?;
+                let scan_file = CDFScanFile {
+                    scan_type: CDFScanFileType::Cdc,
+                    path,
+                    dv_info: DvInfo {
+                        deletion_vector: None,
+                    },
```

Comment:
nit: net fewer lines to hoist `dv_info` out as its own variable (like the other two did)